server.go

```diff
@@ -1607,104 +1566,83 @@ func (s *Server) WriteSeries(database, retentionPolicy string, points []Point) (
 		retentionPolicy = rp.Name
 	}
 
-	// Collect responses for each channel.
-	type resp struct {
-		index uint64
-		err   error
+	// Ensure all required Series and Measurement Fields are created cluster-wide.
+	if err := s.createMeasurementsIfNotExists(database, retentionPolicy, points); err != nil {
+		return 0, err
 	}
-	ch := make(chan resp, len(points))
 
-	// Write each point in parallel.
-	var wg sync.WaitGroup
-	for i := range points {
-		wg.Add(1)
-		go func(p *Point) {
-			index, err := s.writePoint(database, retentionPolicy, p)
-			ch <- resp{index, err}
-			wg.Done()
-		}(&points[i])
+	// Ensure all the required shard groups exist. TODO: this should be done async.
+	if err := s.createShardGroupsIfNotExists(database, retentionPolicy, points); err != nil {
+		return 0, err
 	}
-	wg.Wait()
-	close(ch)
 
-	// Calculate max index and check for errors.
-	var index uint64
-	var err error
-	for resp := range ch {
-		if resp.index > index {
-			index = resp.index
-		}
-		if err == nil && resp.err != nil {
-			err = resp.err
+	// Build writeRawSeriesMessageType publish commands.
+	shardData := make(map[uint64][]byte)
+	for _, p := range points {
+		// Local function makes lock management foolproof.
+		measurement, series, err := func() (*Measurement, *Series, error) {
+			s.mu.RLock()
+			defer s.mu.RUnlock()
+			db := s.databases[database]
+			if db == nil {
+				return nil, nil, fmt.Errorf("database not found %q", database)
+			}
+			if measurement, series := db.MeasurementAndSeries(p.Name, p.Tags); series != nil {
+				return measurement, series, nil
+			}
+			panic("series not found")
+		}()
+		if err != nil {
+			return 0, err
 		}
```

Comment:
The operation of looking up measurement/series should be pretty fast so requiring a lock/unlock on every point seems excessive. I'd wrap the `for` loop in the `func()` and just lock once.